### PR TITLE
8306302: C2 Superword fix: use VectorMaskCmp and VectorBlend instead of CMoveVF/D

### PR DESCRIPTION
**Pull request by eme64**
**Bug**
In `x86`, `CMoveVF/D` were not correctly implemented for the `eq` and `neq` case (leads to assert). And the `lt/le/gt/ge` cases did not all handle `NaN's` correctly (ordered vs unordered comparision, leads to wrong results).

The assert gets triggered in the code from this change: [JDK-8285973](https://bugs.openjdk.org/browse/JDK-8285973)
On this line: https://github.com/openjdk/jdk/commit/c1db70d827f7ac81aa6c6646e2431f672c71c8dc#diff-e5266a3774f26ac663dcc67e0be18608b1735f38c0576673ce36e0cd689bab4aR4309

The problematic line wants to find a Cmp above the Bool, and compare its inputs. But we have no Cmp there, just a constant, that we have set during matching:
https://github.com/openjdk/jdk/blob/af4d5600e37ec6d331e62c5d37491ee97cad5311/src/hotspot/share/opto/matcher.cpp#L2394

The wrong results with `NaN` are because of a bug in `x`:
https://github.com/openjdk/jdk/commit/0485593fbc4a3264b79969de192e8e7d36e5b590#diff-7070c036c7d88ba4a8467e404d8d88aee646b97bf7bacc8b73cbc93f3ef11d2dR2106
The cases `lt` and `le` include the `-1` case, which shoud return `true` if any comparison input is a `NaN`, just as defined for java bytecode `fcmpl/dcmpl`. But they were mapped to ordered comparison codes, not unordered ones. More [here](https://bugs.openjdk.org/browse/JDK-8306302?focusedCommentId=14579078&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14579078).

**Solution**
@fg1417 suggested that `CMoveVF/D` is perfectly composed of `VectorMaskCmp + VectorBlend`. So instead of fixing `CMoveVF/D`, I replaced it. Performance should be the same, as it goes down to the same assembly instructions.

This has a few benefits:
 - `VectorMaskCmp + VectorBlend` is more powerful:
   - `CMoveVF/D` required the same inputs to the compare than to the move itself.
   - `CMoveVF/D` on x86 was only implemented for 32 bytes. Any other size would simply fail to vectorize.
   - `VectorMaskCmp` and `VectorBlend` can have different compare inputs, and even different types. For now, the input types must have the same data-width (`float` and `int`, `double` and `long`).
   - We need less code (I completely removed all code for `CMoveVF/D`).

I also moved the whole `CMove` code in `SuperWord` into `SuperWord::output`, rather than the complex code `SuperWord::merge_packs_to_cmove / CMoveKit`.

As reported in [JDK-8306088](https://bugs.openjdk.org/browse/JDK-8306088) https://github.com/openjdk/jdk/pull/13354, the CMove code did not properly maintain the `packset` / `my_pack`. I now added some verification here, since I also just removed the problematic code, and the verification passes now with this patch. I was also able to remove the unwanted `UseVectorCmov` in an assert.

Most of the changes come from the regression tests in `TestVectorConditionalMove.java`: I generalized it from `aarch64` to all platforms, the IR rules only apply with `avx/asimd`. And I added many new tests to cover the newly implemented cases. Further, I modified the tests to include `NaN's` among the random numbers, to verify that the ordered/unordered comparisions are correct.

**Discussion / Context / Future Work**

1. From what I understand, we currently never introduce a `CMoveF/D`, unless asked for by `UseCMoveUnconditionally` (`C->use_c_move()`). If the flag is set, we attribute no cost to the CMove, else we take `Matcher::float_cmove_cost()`, which seems to be `ConditionalMoveLimit`, and so the Phi is never converted into a CMove.

   An then if one wants to convert these scalar-CMove into a vector-CMove, one needs to activate the flag `UseVectorCmov`. @vnkozlov did some research: the goal was always to have this be on by default eventually.

   I see 2 paths here: either we obsolete `UseVectorCmov`, and implicitly have it on. Or we keep it, but make it by default on. I can do some performance measurements in a follow-up **RFE**.

2. I also saw that `int` and `long` are also CMove'd in `PhaseIdealLoop::conditional_move`. Especially `int` can currently be CMove'd without the `UseCMoveUnconditionally` flag. It would be nice to allow them to be vectorized. This is a small fix, but I'd like to do the testing and performance analysis for it. So a separate **RFE**. A slightly more involved idea: also allow cmp / blend with types of different widths (eg. compare `int` but cmove `double`). That would require a cast on the vector-mask.

3. It is a shame that scalar-CMove is on its own usually not profitable. But together with `SuperWord` it would be profitable. But if it is not scalar-CMove'd first, we fail to vectorize, since the loop has control-flow. It is one of my dreams: allow `SuperWord` to handle control flow, and to the `If-conversion` (CMove) directly with vectorization. Let me know if you have any thoughts or ideas.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306302](https://bugs.openjdk.org/browse/JDK-8306302): C2 Superword fix: use VectorMaskCmp and VectorBlend instead of CMoveVF/D

### Reviewers
 * [Fei Gao](https://openjdk.org/census#fgao) (@fg1417 - Committer)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13493/head:pull/13493` \
`$ git checkout pull/13493`

Update a local copy of the PR: \
`$ git checkout pull/13493` \
`$ git pull https://git.openjdk.org/jdk.git pull/13493/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13493`

View PR using the GUI difftool: \
`$ git pr show -t 13493`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13493.diff">https://git.openjdk.org/jdk/pull/13493.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13493#issuecomment-1539499376)

### CODE REVIEW

**Comment by eme64** [comment]
@fg1417 Since I'm basically implementing your suggestion: do you agree with this fix?

File: src/hotspot/share/opto/superword.cpp
Comment (fg1417) [inline src/hotspot/share/opto/superword.cpp:2855]: Sorry, I'm a bit confusing about the comment here. Based on your following description, are `LT` and `LE` unordered either?

**Comment by fg1417** [comment]
> @fg1417 Since I'm basically implementing your suggestion: do you agree with this fix?

Hi @eme64 , nice rewrite!

BTW, have you tested your patch with `-XX:+UseCMoveUnconditionally` and `-XX:+UseVectorCmov` for all jtreg? Thanks.

File: src/hotspot/share/opto/superword.cpp
Comment (eme64) [reply src/hotspot/share/opto/superword.cpp:2855]: The issue is this:
`CmpF -> Bool [lt/le]` is unordered, because they both accept the return code `-1` from the `CmpF`, which also makes comparisons with `NaN` true. This means that such comparisons are `unordered`.

But `VectorMaskCmp` would interpret `lt/le` test-codes as `ordered`, so they would return false for `NaN` comparisons. So that is why we need to make a transformation here.

How would you improve my comments?

Comment (eme64) [reply src/hotspot/share/opto/superword.cpp:2855]: @fg1417 thanks for the suggestion about running with the flags over all jtreg. I'll do that now...

Comment (fg1417) [reply src/hotspot/share/opto/superword.cpp:2855]: > The issue is this: `CmpF -> Bool [lt/le]` is unordered, because they both accept the return code `-1` from the `CmpF`, which also makes comparisons with `NaN` true. This means that such comparisons are `unordered`.
> 
> But `VectorMaskCmp` would interpret `lt/le` test-codes as `ordered`, so they would return false for `NaN` comparisons. So that is why we need to make a transformation here.
> 
> How would you improve my comments?

Thanks for your clarification.

Your comment is quite clear already. Maybe just highlight the mismatch between `VectorMaskCmp` and `bol_test` here, like:
```
          //
          // But with these two cases, which `VectorMaskCmp` interprets as ordered,
          // we must convert the unordered into an ordered comparison:
          //   BoolTest::lt:  Case -1    -> LT_U
          //   BoolTest::le:  Case -1, 0 -> LE_U
          //
```

Comment (eme64) [reply src/hotspot/share/opto/superword.cpp:2855]: @fg1417 Are you ok with how I worded it now?

Comment (fg1417) [reply src/hotspot/share/opto/superword.cpp:2855]: > @fg1417 Are you ok with how I worded it now?

Oh, yes. Clear enough!

**Comment by eme64** [comment]
@jatin-bhateja or @sviswa7 would you mind reviewing this too?

File: src/hotspot/share/opto/superword.cpp
Comment (eme64) [reply src/hotspot/share/opto/superword.cpp:2855]: @fg1417 Is there anything you still want me to change before you could approve this PR?

**Review by fg1417 — APPROVED** [review]
Nice work! Mid-end part and aarch64 part are generally good to me. Maybe you need another review from an expert on x86.

File: src/hotspot/share/opto/superword.cpp
Comment (fg1417) [reply src/hotspot/share/opto/superword.cpp:2855]: > @fg1417 Is there anything you still want me to change before you could approve this PR?

Sorry for my delay. I assume that the flags `-XX:+UseCMoveUnconditionally` and `-XX:+UseVectorCmov` over all jtreg passed.

Comment (eme64) [reply src/hotspot/share/opto/superword.cpp:2855]: Ah yes, they have passed indeed. With and without those flags, up to tier5 and stress testing :blush:

**Comment by eme64** [comment]
Thanks @fg1417 for the review!

Yes, the testing passes up to at least tier5 and stress testing. With and without `-XX:+UseCMoveUnconditionally` and `-XX:+UseVectorCmov`

Yes, I hope that someone from intel / x86 specialists can review this too :)

**Comment by sviswa7** [comment]
@eme64 I will take a look at it next week.

**Comment by jatin-bhateja** [comment]
> Thanks @fg1417 for the review!
> 
> Yes, the testing passes up to at least tier5 and stress testing. With and without `-XX:+UseCMoveUnconditionally` and `-XX:+UseVectorCmov`
> 
> Yes, I hope that someone from intel / x86 specialists can review this too :) These are candidates: @jatin-bhateja @sviswa7 @merykitty

Your patch looks good to me.  

Patch testing with UseSSE=2 shows failure in following tests, failure is unrelated to your changes, we need to add a strict feature based check in test tag  * @requires os.simpleArch == "x64" & (vm.cpu.features ~= ".*avx.*") or use applyIfCPUFeature

make test TEST="test/hotspot/jtreg/compiler/intrinsics/math/TestFpMinMaxReductions.java" CONF=linux-x86_64-server-fastdebug JTREG="RETAIN=all;JOBS=8;TIMEOUT_FACTOR=8;JAVA_OPTIONS=-XX:UseSSE=2"

**Review by jatin-bhateja — APPROVED** [review]

**Comment by eme64** [comment]
@jatin-bhateja I could reproduce it on master, so it is indeed unrelated.
Filed the bug https://bugs.openjdk.org/browse/JDK-8308746

Thanks @jatin-bhateja @fg1417 thanks for the reviews!
/integrate

**Pull request closed without merging**
